util/dvsim/FlowCfg.py

```diff
@@ -153,94 +156,103 @@ def check_if_primary_cfg(self, hjson_dict):
+        '''Set an attribute from an hjson file
+
+        The path argument is the path for the hjson file that we're reading.
+
+        '''
+        # Is this value overridden on the command line? If so, use the override
+        # instead.
+        args_val = None
+        if key in _CMDLINE_FIELDS:
+            args_val = getattr(self.args, key, None)
+        override_msg = ''
+        if args_val is not None:
+            dict_val = args_val
+            override_msg = ' from command-line override'
+
+        self_val = getattr(self, key, None)
+        if self_val is None:
+            # A new attribute (or the old value was None, in which case it's
+            # just a placeholder and needs writing). Set it and return.
+            setattr(self, key, dict_val)
+            return
+
+        # This is already an attribute. Firstly, we need to make sure the types
+        # are compatible.
+        if type(dict_val) != type(self_val):
+            log.error("Conflicting types for key {!r} when loading {!r}. "
+                      "Cannot override value {!r} with {!r} (which is of "
+                      "type {}{})."
+                      .format(key, path, self_val, dict_val,
+                              type(dict_val).__name__, override_msg))
+            sys.exit(1)
+
+        # Looks like the types are compatible. If they are lists, concatenate
+        # them.
+        if isinstance(self_val, list):
+            setattr(self, key, self_val + dict_val)
+            return
+
+        # Otherwise, check whether this is a type we know how to deal with.
+        scalar_types = {str: [""], int: [0, -1], bool: [False]}
+
+        defaults = scalar_types.get(type(dict_val))
+        if defaults is None:
+            log.error("When loading {!r} and setting key {!r}, found a value "
+                      "of {!r}{} with unsupported type {}."
+                      .format(path, key, dict_val,
+                              override_msg, type(dict_val).__name__))
+            sys.exit(1)
+
+        # If the values are equal, there's nothing more to do
+        if self_val == dict_val:
+            return
+
+        old_is_default = self_val in defaults
+        new_is_default = dict_val in defaults
+
+        # Similarly, if new value looks like a default, ignore it (regardless
+        # of whether the current value looks like a default).
+        if new_is_default:
+            return
+
+        # If the existing value looks like a default and the new value doesn't,
+        # take the new value.
+        if old_is_default:
+            setattr(self, key, dict_val)
+            return
+
+        # Neither value looks like a default. Raise an error.
+        log.error("When loading {!r}, key {!r} is given a value of "
+                  "{!r}{}, but the key is already set to {!r}."
+                  .format(path, key, dict_val, override_msg, self_val))
+        sys.exit(1)
+
+    def resolve_hjson_raw(self, path, hjson_dict):
```

Comment:
The newly added `path` argument (to this and to `_set_attribute` methods) serves no other purpose than to make the error messages clear. Is that correct?

Reply:
Yes, that's right. Since we have configuration files that can include each other (and the hjson library doesn't annotate its parsed data with line information) I think the very least we can do is tell the poor user which of their files has an error :-)